### PR TITLE
Add functionality to role.py and refactor cluster_role.py

### PR DESCRIPTION
**Pull request by Anatw**
##### Short description:
 - Add functionality to role.py.
 - Refactor cluster_role.py.
 - Add add_rules() functionality.
 - Add a common function to utils.py
 - Add an example to the examples folder.

### CODE REVIEW

**Comment by Anatw** [comment]
!verified

File: ocp_resources/role.py [hunk 1]
```diff
@@ -9,3 +9,62 @@ class Role(NamespacedResource):
...
+        self,
+        name=None,
+        namespace=None,
+        client=None,
+        api_groups=None,
+        permissions_to_resources=None,
+        verbs=None,
+        teardown=True,
+        yaml_file=None,
+        delete_timeout=TIMEOUT_4MINUTES,
```
Comment (myakove) [inline ocp_resources/role.py:23]: Please add docstring

File: ocp_resources/role.py [hunk 2]
```diff
@@ -9,3 +9,62 @@ class Role(NamespacedResource):
...
+                verbs=self.verbs,
+            )
+
+    def add_rule(
+        self,
+        verbs,
+        api_groups=None,
+        permissions_to_resources=None,
+    ):
+        if not self.res:
```
Comment (myakove) [inline ocp_resources/role.py:57]: `if not self.res:` this is not needed, you check this in `to_dict()`

File: ocp_resources/role.py [hunk 3]
```diff
@@ -9,3 +9,62 @@ class Role(NamespacedResource):
...
+            super().to_dict()
+
+        rule = {"verbs": verbs}
+        if api_groups:
+            rule["apiGroups"] = api_groups
+        if permissions_to_resources:
+            rule["resources"] = permissions_to_resources
+        if rule:
```
Comment (myakove) [inline ocp_resources/role.py:65]: `if rule:` will always be `True`, you set it in line 60

**Review by myakove — CHANGES_REQUESTED** [review]
Rename the PR title, it's wrong and do not describe the PR content.

File: ocp_resources/role.py [hunk 4]
```diff
@@ -9,3 +9,62 @@ class Role(NamespacedResource):
...
+
+    def to_dict(self):
+        if not self.res:
+            super().to_dict()
+
+        if not self.yaml_file and self.permissions_to_resources:
+            self.add_rule(
+                api_groups=self.api_groups,
+                permissions_to_resources=self.permissions_to_resources,
+                verbs=self.verbs,
```
Comment (rnetser) [inline ocp_resources/role.py:48]: does `verbs=None` a valid use case?

**Review by rnetser — CHANGES_REQUESTED** [review]
Please update the description and title - to match what is done is the RP

File: ocp_resources/role.py (hunk 2 above)
Comment (Anatw) [reply ocp_resources/role.py:57]: There are scenarios in which add_rule will be called before to_dict. It is needed in both functions.

File: ocp_resources/role.py (hunk 3 above)
Comment (Anatw) [reply ocp_resources/role.py:65]: ack

File: ocp_resources/role.py (hunk 4 above)
Comment (Anatw) [reply ocp_resources/role.py:48]: No. I'll add a default value.

File: ocp_resources/role.py (hunk 1 above)
Comment (Anatw) [reply ocp_resources/role.py:23]: ack.

**Comment by Anatw** [comment]
!verified

File: ocp_resources/cluster_role.py [hunk 5]
```diff
@@ -16,9 +16,9 @@ def __init__(
         client=None,
         api_groups=None,
         permissions_to_resources=None,
-        verbs=None,
         teardown=True,
         yaml_file=None,
+        verbs=[""],
```
Comment (rnetser) [inline ocp_resources/cluster_role.py:21]: Do not use mutable default arguments in Python

File: ocp_resources/role.py [hunk 6]
```diff
@@ -9,3 +9,66 @@ class Role(NamespacedResource):
...
+    def __init__(
+        self,
+        name=None,
+        namespace=None,
+        client=None,
+        api_groups=None,
+        permissions_to_resources=None,
+        teardown=True,
+        yaml_file=None,
+        verbs=[""],
```
Comment (rnetser) [inline ocp_resources/role.py:22]: ditto

File: ocp_resources/role.py [hunk 7]
```diff
@@ -9,3 +9,66 @@ class Role(NamespacedResource):
     """
 
     api_group = NamespacedResource.ApiGroup.RBAC_AUTHORIZATION_K8S_IO
+
```
Comment (rnetser) [inline ocp_resources/role.py:12]: Please follow https://github.com/RedHatQE/openshift-python-wrapper/blob/main/CONTRIBUTING.md guielines

File: ocp_resources/role.py [hunk 8]
```diff
@@ -9,3 +9,66 @@ class Role(NamespacedResource):
...
+        api_groups=None,
+        permissions_to_resources=None,
+        teardown=True,
+        yaml_file=None,
+        verbs=[""],
+        delete_timeout=TIMEOUT_4MINUTES,
+        **kwargs,
+    ):
+        """
+        Role object, inherited from NamespacedResource.
```
Comment (rnetser) [inline ocp_resources/role.py:27]: Please use proper docstring format

File: ocp_resources/role.py [hunk 9]
```diff
@@ -9,3 +9,66 @@ class Role(NamespacedResource):
...
+            super().to_dict()
+
+        if not self.yaml_file and self.permissions_to_resources:
+            self.add_rule(
+                api_groups=self.api_groups,
+                permissions_to_resources=self.permissions_to_resources,
+                verbs=self.verbs,
+            )
+
+    def add_rule(
```
Comment (rnetser) [inline ocp_resources/role.py:56]: duplicate code

**Review by rnetser — CHANGES_REQUESTED** [review]
Please update the PR title to match the code

File: ocp_resources/role.py (hunk 8 above)
Comment (Anatw) [reply ocp_resources/role.py:27]: ack.

File: ocp_resources/role.py (hunk 6 above)
Comment (Anatw) [reply ocp_resources/role.py:22]: ack.

File: ocp_resources/role.py (hunk 7 above)
Comment (Anatw) [reply ocp_resources/role.py:12]: ack.

File: ocp_resources/role.py (hunk 2 above)
Comment (myakove) [reply ocp_resources/role.py:62]: Please add an example to the README on how to add multiple rules

File: ocp_resources/role.py (hunk 8 above)
Comment (myakove) [reply ocp_resources/role.py:27]: Missing `Args:` in the docstring

**Comment by Anatw** [comment]
!verified

File: README.md [hunk 10]
```diff
@@ -152,6 +152,33 @@ with NodeNetworkConfigurationPolicy(
...
+latency_role.add_rule(
+    api_groups=["subresources.kubevirt.io"],
+    permissions_to_resources=["virtualmachineinstances/console"],
+    verbs=["get"],
+)
+latency_role.add_rule(
+    api_groups=["k8s.cni.cncf.io"],
+    permissions_to_resources=["network-attachment-definitions"],
+    verbs=["get"],
+)
```
Comment (rnetser) [inline README.md:176]: Should have `add_rules` which gets a list of dicts with rules.
 `add_rules` should call  `add_rule` which should add one rule at a time.
The idea is to be able to create a resource with multiple rules (should be updated in the code in all places)

**Review by rnetser — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (Anatw) [reply README.md:176]: Ack.

**Comment by Anatw** [comment]
!verified

**Comment by Anatw** [comment]
!verified

File: ocp_resources/role.py (hunk 2 above)
Comment (Anatw) [reply ocp_resources/role.py]: ack.

File: ocp_resources/cluster_role.py [hunk 11]
```diff
@@ -30,39 +29,12 @@ def __init__(
             delete_timeout=delete_timeout,
             **kwargs,
         )
-        self.api_groups = api_groups
-        self.permissions_to_resources = permissions_to_resources
-        self.verbs = verbs
+        self.rules = rules
```
Comment (myakove) [inline ocp_resources/cluster_role.py:32]: I guess rules arg is mandatory, please raise if not passed by the user (and not yaml_file)

File: ocp_resources/utils.py [hunk 12]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
...
+
+def add_rules(role_object):
+    """
+    Add a list of rules to a Role / CluterRole object.
+
+    Args:
+        role_object
+    """
+
+    def add_rule(
```
Comment (myakove) [inline ocp_resources/utils.py:289]: please rename to `_add_rule`

**Review by myakove — CHANGES_REQUESTED** [review]

File: examples/role.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    namespace="namespace_name",
+    rules=rules,
+)
+latency_role.deploy()
+
+# delete Role
+latency_role.clean_up()
+
+# Instead of directly calling deploy() and clean_up(), you can use the ``with`` statement that both deploy the
+# resource and ensures automatic clean-up of the code executed.
```
Comment (rnetser) [inline examples/role.py:34]: Already exists in `examples/general.py` no need to repeat

File: ocp_resources/utils.py [hunk 14]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
...
                     _msg=skip_create_warn_msg,
                 )
+
+
+def add_rules(role_object):
+    """
+    Add a list of rules to a Role / CluterRole object.
+
+    Args:
+        role_object
```
Comment (rnetser) [inline ocp_resources/utils.py:286]: Please use proper docstring format

File: ocp_resources/utils.py [hunk 15]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
...
+
+    def add_rule(
+        _role_object,
+        verbs,
+        api_groups,
+        permissions_to_resources,
+    ):
+        def _set_rule(rule_to_set):
+            _role_object.desired_state["rules"].append(rule_to_set)
+            _role_object.res["rules"] = _role_object.desired_state["rules"]
```
Comment (rnetser) [inline ocp_resources/utils.py:297]: any reason why this requires a separate function?

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role.py (hunk 11 above)
Comment (Anatw) [reply ocp_resources/cluster_role.py:32]: ack.

File: ocp_resources/utils.py (hunk 12 above)
Comment (Anatw) [reply ocp_resources/utils.py:289]: Ack.

File: examples/role.py (hunk 13 above)
Comment (Anatw) [reply examples/role.py:34]: Ack.

File: ocp_resources/utils.py (hunk 14 above)
Comment (Anatw) [reply ocp_resources/utils.py:286]: Ack.

File: ocp_resources/utils.py (hunk 15 above)
Comment (Anatw) [reply ocp_resources/utils.py:300]: I will move it to the loop in add_rules()

File: ocp_resources/cluster_role.py [hunk 16]
```diff
@@ -30,39 +33,14 @@ def __init__(
...
-            rule["apiGroups"] = api_groups
-        if permissions_to_resources:
-            rule["resources"] = permissions_to_resources
-        if rule:
-            self._set_rule(rule=rule)
-
-    def _set_rule(self, rule):
-        self.desired_state["rules"].append(rule)
-        self.res["rules"] = self.desired_state["rules"]
+        if not self.yaml_file and self.rules:
```
Comment (myakove) [inline ocp_resources/cluster_role.py:45]: no need to check `self.rules` here, you will have it if you got here .

File: ocp_resources/role.py [hunk 17]
```diff
@@ -9,3 +12,39 @@ class Role(NamespacedResource):
...
+        if not rules and not yaml_file:
+            raise ValueError("must send rules or yaml_file")
+        self.rules = rules
+        self.desired_state = {"rules": []}
+
+    def to_dict(self):
+        if not self.res:
+            super().to_dict()
+
+        if not self.yaml_file and self.rules:
```
Comment (myakove) [inline ocp_resources/role.py:49]: `and self.rules:` ditto

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/cluster_role.py [hunk 18]
```diff
@@ -30,39 +33,14 @@ def __init__(
             delete_timeout=delete_timeout,
             **kwargs,
         )
-        self.api_groups = api_groups
-        self.permissions_to_resources = permissions_to_resources
-        self.verbs = verbs
+        if not rules and not yaml_file:
```
[the remark this answers is not in the document]
Comment (myakove) [reply ocp_resources/cluster_role.py:36]: Also I think this should only be check in to_dict()

User can initialize a class instance and after set rules and call deploy()

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/utils.py (hunk 15 above)
Comment (Anatw) [reply ocp_resources/utils.py:300]: I meant I'll move the content of it into the other function, so we will only have the add_rules().

File: ocp_resources/cluster_role.py (hunk 16 above)
Comment (Anatw) [reply ocp_resources/cluster_role.py:45]: Ack.

File: ocp_resources/cluster_role.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/cluster_role.py:36]: If I set it to "or" I'm demanding the user to add both roles and yaml_file. Using the |and" allow the user to only pass one of them.

[the remark this answers is not in the document]
Comment (Anatw) [reply ocp_resources/cluster_role.py:36]: Agreed, moving it to the to_dict() method.

File: ocp_resources/utils.py [hunk 19]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
...
                     _msg=skip_create_warn_msg,
                 )
+
+
+def add_rules(role_object):
+    """
+    Add a list of rules to a Role / CluterRole object.
+
+    Args:
+        role_object (list): list of dicts of rules. In the dict:
```
Comment (myakove) [inline ocp_resources/utils.py:286]: `role_object (list)` is not a list, it's a resource

File: ocp_resources/utils.py [hunk 20]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
...
+                permissions_to_resources (list): List of string with resource/s to which you want to add permissions to.
+                Verbs (list): Determine the action/s (permissions) applicable on a specific resource.
+                    Available verbs per resource can be seen with the command 'oc api-resources --sort-by name -o wide'
+    """
+
+    def _add_rule(
```
Comment (myakove) [inline ocp_resources/utils.py:292]: The inner function is not needed, 
can be called in the for loop

File: ocp_resources/utils.py [hunk 21]
```diff
@@ -276,3 +276,39 @@ def _return_resource(_resource, _check_exists, _msg):
                     _check_exists=check_exists,
                     _msg=skip_create_warn_msg,
                 )
+
+
+def add_rules(role_object):
```
Comment (myakove) [inline ocp_resources/utils.py:281]: I think this function should generate the rules and return then, not set them in the resource.
Should get `role_object.rules` and return the `_role_object.desired_state["rules"]`

Then in the resource you need to set it
`self.res["rules"] = _role_object.desired_state["rules"]`

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/utils.py (hunk 19 above)
Comment (Anatw) [reply ocp_resources/utils.py:286]: Ack.

File: ocp_resources/utils.py (hunk 20 above)
Comment (Anatw) [reply ocp_resources/utils.py:292]: Ack.

**Comment by Anatw** [comment]
!verified

File: ocp_resources/utils.py [hunk 22]
```diff
@@ -276,3 +276,25 @@ def _return_resource(_resource, _check_exists, _msg):
                     _check_exists=check_exists,
                     _msg=skip_create_warn_msg,
                 )
+
+
+def generate_rules_list_of_dicts(roles):
```
Comment (myakove) [inline ocp_resources/utils.py:281]: Should be renamed to `generate_rules`

File: ocp_resources/utils.py [hunk 23]
```diff
@@ -276,3 +276,25 @@ def _return_resource(_resource, _check_exists, _msg):
...
+    """
+    Generate a list of dicts of rules.
+
+    Args:
+        roles (list): list of dicts of rules. In the dict:
+                permissions_to_resources (list): List of string with resource/s to which you want to add permissions to.
+                Verbs (list): Determine the action/s (permissions) applicable on a specific resource.
+                    Available verbs per resource can be seen with the command 'oc api-resources --sort-by name -o wide'
+    """
```
Comment (myakove) [inline ocp_resources/utils.py:290]: Missing `Returns` in the docstring

File: ocp_resources/role.py [hunk 24]
```diff
@@ -9,3 +12,37 @@ class Role(NamespacedResource):
...
+        client=None,
+        rules=None,
+        teardown=True,
+        yaml_file=None,
+        delete_timeout=TIMEOUT_4MINUTES,
+        **kwargs,
+    ):
+        """
+        Args:
+            rules (list): list of dicts of rules. More info can be seen in the generate_rules_list_of_dicts function.
```
Comment (myakove) [inline ocp_resources/role.py:29]: Missing many args in docstring

File: ocp_resources/cluster_role.py [hunk 25]
```diff
@@ -14,14 +15,16 @@ def __init__(
...
-        permissions_to_resources=None,
-        verbs=None,
+        rules=None,
         teardown=True,
         yaml_file=None,
         delete_timeout=TIMEOUT_4MINUTES,
         **kwargs,
     ):
+        """
+        Args:
```
Comment (myakove) [inline ocp_resources/cluster_role.py:25]: Missing many args in docstring

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/utils.py (hunk 22 above)
Comment (Anatw) [reply ocp_resources/utils.py:281]: Ack.

File: ocp_resources/cluster_role.py (hunk 25 above)
Comment (Anatw) [reply ocp_resources/cluster_role.py:25]: Ack.

File: ocp_resources/utils.py (hunk 23 above)
Comment (Anatw) [reply ocp_resources/utils.py:290]: Removing the function.

**Comment by Anatw** [comment]
!verified

**Comment by Anatw** [comment]
!verified

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Comment by Anatw** [comment]
!cherry-pick v4.12

**Pull request merged**